Add xpu legacy profiler for Intel(R) XPU devices

xunsongh opened the pull request:
# Motivation
This patch aimed to add an xpu path of legacy profiler for Intel(R) XPU devices. With this PR,  users will be able to profile GPU kernel time consuming and memory allocations out-of-queue while using Intel(R) XPU backend.

# Solution
We add use_xpu for entering xpu path of legacy profiler and we imitate CUDA to designed how xpu_time_total and self_xpu_time_total should be calculated. Different from CUDA and CPU who use 'push' range and 'pop' range to get the interval of an operator, we designed to use 'mark' to mark the XPU kernels to the profiler. One XPU kernel will calculate its elapsed time by the SYCL event related to the operator itself.
Also, we redefined the struct of KernelEventBase, which can be used as a template for all backends to design their own profiler in a unified way, not only for CUDA or XPU, but also for other backends which will raise independent kernels while running and profiling.

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

chengjunlu commented on torch/autograd/profiler_util.py:
```diff
@@ -487,15 +506,47 @@ def cpu_time_total(self):
+            if not self.is_legacy:
+                # account for the kernels in the children ops
+                if len(self.kernels) > 0:
+                    assert self.kernels[0].device.type == DeviceType.CUDA
```
DeviceType.XPU?

chengjunlu commented on torch/csrc/autograd/profiler_legacy.h:
```diff
@@ -209,13 +274,24 @@ struct TORCH_API LegacyEvent {
 
   bool hasCuda() const {
-    return cuda_event != nullptr || (isRemote() && device_ != -1);
+    return hasKernel() && device_.type() == DeviceType::CUDA;
```
May the device type be CPU and the kernel is for the CUDA?

xunsongh committed:
To support legacy profiler on Intel(R) XPU devices

This patch will add a path into legacy profiler for profiling Intel(R)
  XPU kernels. We also defined KernelEventBase and KernelEventStub for
  all GPU kernels including CUDA and XPU, this will be a template class
  for usage or further design.

Signed-off-by: Xunsong, Huang <xunsong.huang@intel.com>

gujinghui commented:
@ezyang  Could you help review this PR? Thanks.

gujinghui commented:
Hi @ezyang

Could you help review this PR?

Thanks.

ezyang commented:
I suppose it has to be asked... why are we adding new functionality to the legacy profiler?

gujinghui commented:
> I suppose it has to be asked... why are we adding new functionality to the legacy profiler?

Hi @ezyang ,

We are going to support both legacy profiler and kineto.
This PR is only for legacy profiler. And we will plan to support kineto in next step.

The motivation of supporting legacy is, we still have many users who are using legacy profiler, both internal and external.
Therefore, we need to support it, as well.

Thanks,

gdankel commented:
@gujinghui as a general rule we won't accept patches to the legacy profiler. Can you elaborate on why people are still using the legacy profiler? I believe we have feature parity with Kineto. If there are any other reasons please let us know and we will address them. We do not want to continue development on the legacy profiler.

ezyang commented:
(unsubscribing myself; @gdankel can you drive this to conclusion?)
